libs/README.md

```diff
@@ -0,0 +1,375 @@
+```
+
+All the methods above, in most cases, are adapted methods of Docker.
+For example, the `pull_image` method pulls an image with a specific name,
+and `start_service` starts a particular container. To get events from containers,
+we use the `events` method.
+
+A vast number of errors appear because the frontend part has not considered
+all the exceptional cases: for example, when grpc is not started yet, but
+the frontend tried to connect to it. Or if the **Tor** container is not started,
+but the frontend tried to launch the **Base Node**, that failed with an error
+because it couldn't connect to the **Tor** server.
+
+For the last problem, we have a naive solution that is not solved
+the problem entirely: we added a `WAIT_FOR_TOR` environment variable
+in the **Base Node**, but I often see cases when the tor runs longer
+than that interval expected.
+
+The obvious solution to this problem is: let's just take into account
+all these cases in the frontend part. But how to test all that? To cover
+that cases with tests, we have to create a lot of fragile tests that will
+break simply because we move the buttons to other places.
+
+Another potential problem is the destruction of the main window since
+the frontend part controls the backend. If the main window is destroyed
+accidentally by the system, we lose the state, and the containers
+become unmanaged.
+
+It is also impossible to create a multi-window interface, or
+a container that manages others (a kind of launchpad for servers).
+
+Another class of potential problems is related to the fact that we,
+for example, send SQL queries directly from the frontend to the backend.
+And, in theory, we have to cover that requests with tests as well.
+
+### The frontend
+
+The frontend uses `React` framework and `Redux` library.
+
+The `Redux` maintains the main state of the app and calls API methods
+that we exported from the backend.
+
+## The improved architecture
+
+The approach discussed below requires a lot of effort, but it can
+significantly increase the reliability of the launchpad application.
+
+The idea is to move all control functions to the backend and transfer
+the application state and changes to this state to the frontend using
+Tauri events.
+
+In this case, in general, you will not need to export your API methods
+since the application will also be controlled by sending messages from
+frontend to backend, which will launch processes that change that
+mentioned above - shared state.
+
+### Events channels
+
+To connect UI to the new engine we should use two calls on the frontend that is
+the part of the Tauri API.
+
+To subscribe to the state and changes of that we should subscribe to the
+`tari://reactions` stream of events:
+
+```js
+const unlisten = await listen('tari://reactions', (event) => { })
+```
+
+And to send a control action we should send an event to the `tari://actions`
+stream:
+
+```js
+function sendEvent(action) {
+    emit('tari://actions', action)
+}
+```
+
```

Comment:
So you're suggesting that the FE must become a Finite State Machine that reacts to events from the `reactions` stream, and can request state changes via the `actions` stream?